llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp

```diff
@@ -1532,6 +1532,41 @@ static Instruction *foldBitOrderCrossLogicOp(Value *V,
   return nullptr;
 }
 
+static Value *foldBinaryIntrinsicRecurrence(InstCombinerImpl &IC,
+                                            IntrinsicInst *II) {
+  PHINode *PN;
+  Value *Init, *OtherOp;
+
+  // A binary intrinsic recurrence with loop-invariant operands is equivalent to
+  // `call @llvm.binary.intrinsic(Init, OtherOp)`.
+  if (!matchSimpleBinaryIntrinsicRecurrence(II, PN, Init, OtherOp) ||
+      !IC.getDominatorTree().dominates(OtherOp, PN))
+    return nullptr;
+
+  auto IID = II->getIntrinsicID();
+  switch (IID) {
+  case Intrinsic::maxnum:
+  case Intrinsic::minnum:
+  case Intrinsic::maximum:
+  case Intrinsic::minimum:
+  case Intrinsic::maximumnum:
+  case Intrinsic::minimumnum:
+  case Intrinsic::smax:
+  case Intrinsic::smin:
+  case Intrinsic::umax:
+  case Intrinsic::umin:
+    break;
+  default:
+    return nullptr;
+  }
+
+  auto *InvariantBinaryInst =
+      IC.Builder.CreateBinaryIntrinsic(IID, Init, OtherOp);
+  if (isa<FPMathOperator>(InvariantBinaryInst))
+    cast<Instruction>(InvariantBinaryInst)->copyFastMathFlags(II);
```

Comment:
```suggestion
  auto *InvariantBinaryInst =
      IC.Builder.CreateBinaryIntrinsic(IID, Init, OtherOp, /*FMFSource=*/II);
```

Reply:
That's unfortunately not possible per assertion `isa<FPMathOperator>(this) && "getting fast-math flag on invalid op"'`  on non-fast-math binary intrinsics. Perhaps we may want to change this in IRBuilder intrinsic creation in the future.

Reply:
Use `dyn_cast<FPMathOperator>(II)`?

Reply:
In this case, I think you'd need another `dyn_cast<Instruction>(InvariantBinaryInst)` to check that `CreateBinaryIntrinsic` didn't constant-fold, as copyFastMathFlags is private to FPMathOperator. Think the current one may be cleaner.